### ISSUE TITLE
CORE-4 — Vendored reduced Clebsch-Gordan basis (triple oracle)

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1558]
**Depends on:** INF-1 (#1550) · **Blocks:** BKD-1 (#1560), ARCH-2 (#1561)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** The central mathematical work and technical risk #1 of the rewrite, and the ticket that removes an architectural defect. Today the reduced symmetric-tensor-product basis is reachable **only** through `cue.reduced_symmetric_tensor_product_basis`: `mace/modules/wrapper_ops.py:428` does `use_reduced_cg = use_reduced_cg and CUET_AVAILABLE`, and `mace/tools/cg.py:123` takes the cueq branch only `if use_cueq_cg and CUET_AVAILABLE`, otherwise falling through to the redundant `_wigner_nj` full basis (`:128-131`) without raising. Measured today at `3b43d63f`, that is **29 parameters versus 86** for `irreps_in=0e+1o+2e+3o`, `correlation=3`, `keep_ir=0e+1o` — per output, 13 vs 28 (`0e`), 16 vs 58 (`1o`), 20 vs 73 (`2e`), with the ν=3 terms carrying the whole gap (8 vs 23, 12 vs 51, 14 vs 65). So a CPU- or AMD-only host trains a **different, over-parametrized network for the same hyperparameters**, and an explicit `--use_reduced_cg True` is downgraded with no warning. Two further facts about the current state, both load-bearing: the CLI flag defaults to **False** (`mace/tools/arg_parser.py:239-244`) while the model classes default to `True` (`mace/modules/models.py:67`), so a CLI-trained model uses the full basis today unless asked otherwise; and the environment override `MACE_USE_CUEQ_CG` (`mace/tools/cg.py:23`, default `"0"`) only takes effect when the caller passes `use_cueq_cg=None`, which the model path never does. **The reduced basis is not cueq-exclusive**: legacy e3nn's `o3.ReducedTensorProducts`, given a fully permutation-symmetric formula, reproduces cueq's reduced path counts — re-verified at `3b43d63f` on outputs `0e`/`1o`/`2e` at ν=2 and ν=3 including the hard cases 51→12 and 65→14, and previously across 21 cases spanning `irreps_in` up to `0e+1o+2e+3o` and correlation 1–3 — so neither basis is more portable than the other, and `wrapper_ops.py:428` / `mace/tools/cg.py:119-126` are bugs, not constraints. The numerics-critical work here is therefore **pinning the path order and the per-path normalization** — that is the v1 on-disk weight format — and golden-locking the **tensors**, not merely their counts: equal counts do not make two bases equal. Span equality against cueq is already verified (column spaces compared on shared random inputs) including `l > 0` outputs (`0e`, `1o`, `2e`; ν = 2 and 3), and `correlation=4` agrees on the cases tested (3=3, 2=2, 9=9) — the golden lock here extends that to the full production grid. `mace_core` computes the basis from first principles (Clebsch-Gordan / Wigner recoupling) with **no e3nn and no cuequivariance** — both bases are elementary group theory, not a library monopoly; use the small MIT-licensed legacy routine as a porting reference only, and drop the dependency. Legacy code to read in place: `mace/tools/cg.py` (`U_matrix_real` at `:106`, `compute_U_cueq`), `mace/tools/cg_cueq_tools.py` (`symmetric_contraction_proj`).

**Interface & constraints:**

1. **Public API** — `mace_core.clebsch_gordan.reduced_symmetric_tensor_product_basis(irreps_in, correlation, keep_ir, dtype)`, pure numpy, importing **neither `cuequivariance` nor `e3nn`** (no interim dependency on either, ever). It must produce the same tensor today's legacy path produces with `use_reduced_cg=True` **and cuequivariance installed** — which is the only configuration in which develop produces the reduced basis at all, and is therefore the oracle configuration. Build-time cost only (~1 s at `lmax=3, ν=3` in the legacy path; results are cached), so caching is expected and exotic optimization is not.

2. **One basis, every device.** The basis is model state, never a backend or host property: one value for every device and every backend, with no `CUET_AVAILABLE` branch — or any equivalent import-conditioned behaviour — surviving into `packages/`. The three legacy switches that must have no counterpart in `packages/` are `wrapper_ops.py:428`, `cg.py:123` and the `MACE_USE_CUEQ_CG` env read at `cg.py:23`. Backends may later *declare* support for a basis; they never *choose* one.

3. **The pinned order, normalization and path selection ARE the file format.** Canonical symmetric-contraction weights are a flat `[Z, A, mul]` tensor over the reduced basis, in the path order, the per-path normalization and the path selection this ticket defines — document all three in `mace_core.clebsch_gordan`, and emit a coupling-tree label per surviving path so the selection is recoverable from a file rather than only reproducible by rerunning the same enumeration. Canonical layout is `mul_ir`. The legacy nested per-`(irrep_out, ν)` weight tensors are contiguous slices of the flat tensor in that path order, so the reference's boundary functions are concatenate/split views (free). The nested and fused parametrizations of the same reduced basis differ by a **block diagonal map with very small blocks**. On most of the measured grid every block is 1×1, so it reduces to a scaled permutation whose scales are square roots of small integers (1, √2, √3, √5, 2, 2√2, 2√3, 3, √14). It is not one everywhere: order and normalization are pinned here, but *which* of the linearly dependent coupling paths survives the reduction is a free choice this module also has to pin, and two enumerations resolving it differently are related by a change of basis inside the degenerate subspaces. On `0e+1o+2e+3o → 0e+1o+2e` at correlation 3 the map is 44 blocks of size one, one of size two and one of size three, worst condition number 2.79. An accelerated backend consumes the canonical form via a map derived once at build time, an index permutation plus a scale vector wherever the blocks are 1×1 and a small dense solve where they are not; nothing about this choice is negotiated per backend.

4. **Golden discipline — tensors, not counts.** The goldens must assert the path order and the scale of each path for fixed `(irreps_in, irreps_out, correlation)`, plus a parameter-count golden per grid point (the number the `CUET_AVAILABLE` branch used to change), and must never be generated from `mace_core.clebsch_gordan` itself — the lock is against an independent derivation.

5. **`full ↔ reduced` conversion rules.** The full basis has rank equal to the reduced parameter count — its extra directions are pure gauge and do not change the function. `full → reduced` is exact and **unique** (measured max error ~1e-13 at fp64 on the published foundation-model cases, 28↔13 and 86↔29 paths); `reduced → full` is exact but under-determined — return the minimum-norm representative — and is **export-only** (DEP-3 (#1595) consumes it; the checkpoint path refuses it). Loading a full-basis artifact performs the exact `full → reduced` conversion **at load time** (explicit, fp64, function-preserving); every persisted v1 checkpoint is reduced. This matters more than it looks: because the CLI default on develop is `--use_reduced_cg False`, the *typical* legacy checkpoint in the wild is a full-basis one, so the load-time conversion is the common path, not the exotic one. The conversion is function-preserving, **not bit-exact in weight space**: gate it on value goldens, never on weight equality.

6. **Bounded-risk escape hatch.** Both conversion directions are exact, so a pinning error discovered after release is a `CanonicalSpec.schema_version` bump with an in-repo migration, not a format break.

7. **Second oracle runs anywhere.** The base `cuequivariance` package is numpy-level descriptor math and imports with no GPU present, so the `allclose` check against `cue.reduced_symmetric_tensor_product_basis` runs on any host — as a **test-only** dependency, never a dependency of `mace_core`. The same is true of the e3nn `o3.ReducedTensorProducts` cross-check, which is what makes the "not cueq-exclusive" claim reproducible on a laptop.

**Task:**
1. `mace_core.clebsch_gordan.reduced_symmetric_tensor_product_basis(irreps_in, correlation, keep_ir, dtype)` in pure numpy, **without importing `cuequivariance` or `e3nn`**, producing the same tensor today's legacy path produces with `use_reduced_cg=True` and cuequivariance present.
2. Make it the single basis for every device and backend — none of `wrapper_ops.py:428`, `cg.py:123` or the `MACE_USE_CUEQ_CG` env switch has a counterpart in `packages/`.
3. `to_canonical`/`from_canonical` as boundary functions (`mul_ir ↔ ir_mul` is a pure reshape once the basis is fixed).
4. Native `full ↔ reduced` conversion (full→reduced exact, reduced→full min-norm), so **loading a full-basis artifact converts it to reduced at load time** (explicit, fp64, function-preserving) — no e3nn, no cueq. `reduced → full` is export-only (DEP-3 (#1595)).

**Out of scope:** any `SegmentedPolynomial` evaluator (not needed — the reference backend keeps its Horner cascade); relitigating the canonical parametrization (settled: reduced basis, flat `[Z, A, mul]`, MACE-pinned order and normalization).

**Acceptance criteria (triple oracle):**
- [ ] Clebsch-Gordan-fragment value-goldens (P0-6) pass at fp64 — asserting the tensors, path order and per-path scale included, not merely path counts.
- [ ] Parameter-count golden per `(irreps_in, irreps_out, correlation)` — the number the `CUET_AVAILABLE` branch used to change — so the device coupling cannot reappear. The grid includes the measured anchor `irreps_in=0e+1o+2e+3o, correlation=3`: 13 (`0e`), 16 (`1o`), 20 (`2e`), 29 for `keep_ir=0e+1o`.
- [ ] `allclose` against `cue.reduced_symmetric_tensor_product_basis` (second oracle; numpy-level, runs on any host with no GPU).
- [ ] `allclose` of the end-to-end contraction against the P0-4 cueq GPU round-trip (third, independent oracle). Phase 1 closeout requires only the two CPU oracles — the goldens and a naive derivation independent of `mace_core.clebsch_gordan`; the GPU confirmation runs best-effort in nightly and becomes blocking on any reduced-basis change.
- [ ] `full → reduced → full` round trip is function-preserving at fp64 on the golden-locked grid (full→reduced exact and unique; reduced→full minimum-norm).
- [ ] No import of `cuequivariance` **or `e3nn`** anywhere in `mace_core`; there is no interim dependency on either.

**Verify:**
```bash
python -m pytest packages/mace-core/tests/test_reduced_clebsch_gordan.py -v
python -m pytest packages/mace-core/tests/test_reduced_clebsch_gordan_cueq_oracle.py -m cueq -v
python -m pytest tests/architecture/test_no_vendor_deps.py -v   # mace_core imports no cuequivariance
```
**Review focus:** the reduced-basis mathematics — path order, per-path normalization, and the full↔reduced maps — double-reviewed by someone with group-theory/e3nn expertise; goldens generated from an independent derivation, never from the code under test; and that the load-time full→reduced path is treated as the common case it is, given develop's `--use_reduced_cg False` default.

---
## Phase 2 — Keystone slice: architecture, backends, data (16 tickets)

> The keystone (ARCH-1 (#1559)→ARCH-4 (#1563) + FM-00 (#1565)) front-loads all the hard bets — two-layer `MACEBackbone`/`MACEOutputs`, compile-first kernel Protocol, no TorchScript on the live path, `PrecisionConfig` — and pins the risk invariants (per-quantity float64 E0/scale-shift, strain-before-edge-vectors, `mul_ir` readout bias) as goldens before capabilities stack on top. **ARCH-5 (#1633)** closes the surface develop already has and the keystone would otherwise silently drop: node-level differentiable inputs (`magmom` → `magforces`), field derivatives (BEC), and the fixed-point re-entrancy the SCF wrapper needs. **PAR-1 (#1572)** is the continuous in-process parity harness that makes the frozen in-tree `mace/` a live differential oracle; once it is green for a capability, that capability's `--engine` default can flip to `v1` with opt-out (the energy/forces/stress opt-in explicitly excludes LAMMPS export and checkpoint until DEP-2 (#1594)).

**Comment by alinelena** [comment]
we shall use clebsch_gordan everywhere rather than cg at the moment we seem to use both...
